File: mobile/welcome-screen/ios-app/ShowUpWelcome/CodeSlotRow.swift
//  CodeSlotRow.swift
//  ShowUp · the six-box code row, shared by phone and email verification
//
//  Mirrors `designsystem/CodeSlotRow.kt`.
//
//  WHAT MAKES THIS ONE COMPONENT AND NOT TWO
//
//  Both screens draw six 49 x 62 boxes at radius 14 with one digit each, and on both the row is
//  decoration for a SINGLE text field behind it rather than six fields. That last part is the
//  reason it is worth sharing: six fields would mean six focus targets, and then backspace, paste
//  and one-time-code autofill each have to be taught to hop between them, which is where per-digit
//  code inputs usually break.
//
//  WHAT THE TWO SCREENS GENUINELY DIFFER ON
//
//  Two things, both from their own spec sheets rather than drift:
//
//   - `halo`. SHOWUP-153's sheet draws a 4pt ring outside the stroke on the active and error
//     slots. SHOWUP-143's does not. Off by default so the older screen keeps what was approved.
//   - `caretBlinks`. 153's sheet specifies a blinking caret and says why: "A static bar reads as a
//     filled slot at a glance — the one thing this row must never be ambiguous about." 143 shipped
//     a static bar. Left as a parameter rather than changed on both, because altering a screen in
//     PO Acceptance is a product call; the concern is recorded here so it can be taken.

import SwiftUI

struct CodeSlotRow: View {
    let digits: String
    /// The whole row turns danger, and no slot is active — the error is the row, not a position.
    var error: Bool = false
    var slotWidth: CGFloat = 49
    var slotHeight: CGFloat = 62
    var gap: CGFloat = 8
    var halo: Bool = false
    var caretBlinks: Bool = false

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        // A TimelineView, not a repeating animation and not a Timer.
        //
        // The spec is `su-caret-blink 1s steps(2, end)` — a HARD on/off, which is not what an
        // interpolating animation produces: `.repeatForever(autoreverses:)` fades, and a fading
        // caret reads as a dimmed digit rather than a cursor. A Timer would work but has to be
        // owned, cancelled and kept off the main actor's way, and `DispatchQueue` is banned by
        // audit/check-swift-concurrency.py. TimelineView is the declarative version: the schedule
        // drives the redraw and the phase is computed, so there is nothing to leak.
        //
        // Paused entirely under Reduce Motion — a blinking element is exactly what that setting
        // is for.
        Group {
            if caretBlinks && !reduceMotion {
                TimelineView(.periodic(from: .now, by: 0.5)) { timeline in
                    row(caretOn: Int(timeline.date.timeIntervalSinceReferenceDate * 2) % 2 == 0)
                }
            } else {
                row(caretOn: true)
            }
        }
    }

    private func row(caretOn: Bool) -> some View {
        HStack(spacing: gap) {
            ForEach(0..<6, id: \.self) { i in
                slot(index: i, caretOn: caretOn)
            }
        }
    }

    private func slot(index i: Int, caretOn: Bool) -> some View {
        let ch: Character? = i < digits.count ? Array(digits)[i] : nil
        let active = !error && i == digits.count && digits.count < 6

        return ZStack {
            RoundedRectangle(cornerRadius: Radius.control)
                .fill(error ? Color.liqDanger.opacity(0.04) : Color.liqElevated)
            RoundedRectangle(cornerRadius: Radius.control)
                .strokeBorder(
                    error ? Color.liqDanger
                        : active ? Color.liqPurple
                        : ch != nil ? Color.liqFg.opacity(0.32)
                        : Color.liqSubtle,
                    lineWidth: 1.5)
            if let ch {
                Text(String(ch))
                    .font(.custom(PS.loraBold, size: 30))
                    .monospacedDigit()
                    .foregroundColor(error ? .liqDangerDigit : .liqFg)
            } else if active {
                Rectangle()
                    .fill(Color.liqPurple)
                    .frame(width: 2, height: halo ? 26 : 28)
                    .opacity(caretOn ? 1 : 0)
            }
        }
        .frame(width: slotWidth, height: slotHeight)
        // The ring is OUTSIDE the stroke, so it is an overlay on a slightly larger shape rather
        // than a shadow: a shadow has a light source and would not be even on all four sides.
        .overlay(
            RoundedRectangle(cornerRadius: Radius.control)
                .strokeBorder(
                    halo ? (error ? Color.liqDanger.opacity(0.10)
                                  : active ? Color.liqPurple.opacity(0.10) : .clear)
                         : .clear,
                    lineWidth: 4)
                .padding(-4)
        )
        .shadow(color: Color(hex: 0x2E0147).opacity(0.04), radius: 2, y: 1)
        .animation(.easeOut(duration: Motion.fast), value: error)
        .animation(.easeOut(duration: Motion.fast), value: digits.count)
    }
}

File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/designsystem/CodeSlotRow.kt
package com.showup.designsystem

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.keyframes
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * The six-box code row, shared by phone verification and email verification.
 *
 * WHAT MAKES THIS ONE COMPONENT AND NOT TWO
 *
 * Both screens draw six 49 x 62 boxes at radius 14 with one digit each, and both are decoration
 * for a SINGLE text field behind them rather than six fields. That last part is the reason the row
 * is worth sharing at all: six fields would mean six focus targets, and then backspace, paste and
 * one-time-code autofill each have to be taught to hop between them, which is where per-digit code
 * inputs usually break. One field holding the whole code makes paste and autofill work for free,
 * and this row is only what that field looks like.
 *
 * WHAT THE TWO SCREENS GENUINELY DIFFER ON
 *
 * Two things, both from their own spec sheets rather than drift:
 *
 *  - [halo]. SHOWUP-153's sheet draws a 4dp ring outside the stroke on the active and error slots.
 *    SHOWUP-143's does not. Off by default so the older screen keeps what was approved.
 *  - [caretBlinks]. 153's sheet specifies a blinking caret and says why: "A static bar reads as a
 *    filled slot at a glance -- the one thing this row must never be ambiguous about." 143 shipped
 *    a static bar. Left as a parameter rather than changed on both, because altering a screen in
 *    PO Acceptance is a product call; the concern is recorded here so it can be taken.
 *
 * The compact tier is a third difference and a smaller one: 143 shrinks to 44 x 56 on a short
 * frame, which is also the floor below which a slot stops being a reliable touch target.
 */
@Composable
fun CodeSlotRow(
    digits: String,
    /** The whole row turns danger, and no slot is active — the error is the row, not a position. */
    error: Boolean = false,
    slotWidth: Dp = 49.dp,
    slotHeight: Dp = 62.dp,
    halo: Boolean = false,
    caretBlinks: Boolean = false,
    modifier: Modifier = Modifier,
) {
    val motion = rememberMotion()
    val ms = if (motion.enabled) Motion.FAST else 0
    val shape = RoundedCornerShape(Radius.control)

    // One transition drives every caret; a per-slot one would restart whenever the active index
    // moved, so the caret would appear to stutter as the user types.
    val blink by rememberInfiniteTransition(label = "caret").animateFloat(
        initialValue = 1f,
        targetValue = 0f,
        // steps(2, end): hard on/off, not a fade. A fading caret reads as a dimmed digit.
        animationSpec = infiniteRepeatable(
            animation = keyframes {
                durationMillis = 1000
                1f at 0
                1f at 499
                0f at 500
                0f at 999
            },
            repeatMode = RepeatMode.Restart,
        ),
        label = "caretBlink",
    )

    Row(
        modifier.fillMaxWidth(),
        // space-between, never weight(1f): at 320 an evenly divided row drops each slot below
        // 44dp and they stop being reliable targets. The gap absorbs the difference instead.
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        repeat(6) { i ->
            val ch = digits.getOrNull(i)
            val active = !error && i == digits.length && digits.length < 6

            val borderColor by animateColorAsState(
                when {
                    error -> Danger
                    active -> Purple
                    ch != null -> Fg.copy(alpha = 0.32f)
                    else -> Border
                },
                tween(ms), label = "slotBorder",
            )
            val haloColor by animateColorAsState(
                when {
                    !halo -> Color.Transparent
                    error -> Danger.copy(alpha = 0.10f)
                    active -> Purple.copy(alpha = 0.10f)
                    else -> Color.Transparent
                },
                tween(ms), label = "slotHalo",
            )
            val fill by animateColorAsState(
                if (error) Danger.copy(alpha = 0.04f) else Elevated,
                tween(ms), label = "slotFill",
            )

            Box(
                Modifier
                    .size(width = slotWidth, height = slotHeight)
                    // The ring is OUTSIDE the stroke, so it is a second border on the same box
                    // rather than a shadow. Modifier.shadow is elevation and has a light source,
                    // so it would not be even on all four sides.
                    .border(4.dp, haloColor, shape)
                    .clip(shape)
                    .background(fill)
                    .border(1.5.dp, borderColor, shape),
                contentAlignment = Alignment.Center,
            ) {
                if (ch != null) {
                    Text(
                        ch.toString(),
                        color = if (error) DangerDigit else Fg,
                        fontFamily = Lora,
                        fontWeight = FontWeight.Bold,
                        fontSize = 30.sp,
                    )
                } else if (active) {
                    Box(
                        Modifier
                            .size(width = 2.dp, height = if (halo) 26.dp else 28.dp)
                            .alpha(if (caretBlinks && motion.enabled) blink else 1f)
                            .background(Purple),
                    )
                }
            }
        }
    }
}
